File: library/src/main/java/com/opensource/svgaplayer/SVGAVideoSpriteFrameEntity.java
package com.opensource.svgaplayer;

import android.graphics.Matrix;
import android.graphics.Path;
import android.text.TextUtils;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.Serializable;

/**
 * Created by cuiminghui on 2016/10/17.
 */
class SVGAVideoSpriteFrameEntity {

    double alpha = 0.0;
    SVGARect layout = new SVGARect();
    Matrix transform = new Matrix();
    Path maskPath = null;
    SVGAVideoShapeEntity[] shapes = new SVGAVideoShapeEntity[0];

    public SVGAVideoSpriteFrameEntity() {}

    SVGAVideoSpriteFrameEntity(JSONObject obj) throws JSONException {
        alpha = obj.optDouble("alpha", 0);
        final JSONObject transformJO = obj.optJSONObject("transform");
        if (transformJO != null) {
            final float[] arr = new float[9];
            double a = transformJO.getDouble("a");
            double b = transformJO.getDouble("b");
            double c = transformJO.getDouble("c");
            double d = transformJO.getDouble("d");
            double tx = transformJO.getDouble("tx");
            double ty = transformJO.getDouble("ty");
            arr[0] = (float) a; // a
            arr[1] = (float) c; // c
            arr[2] = (float) tx; // tx
            arr[3] = (float) b; // b
            arr[4] = (float) d; // d
            arr[5] = (float) ty; // ty
            arr[6] = (float) 0.0;
            arr[7] = (float) 0.0;
            arr[8] = (float) 1.0;
            transform.setValues(arr);
        }
        final JSONObject layoutJO = obj.optJSONObject("layout");
        if (layoutJO != null) {
            double x = layoutJO.getDouble("x");
            double y = layoutJO.getDouble("y");
            double width = layoutJO.getDouble("width");
            double height = layoutJO.getDouble("height");
            layout = new SVGARect(x, y, width, height);
        }
        String clipPath = obj.optString("clipPath");
        if (!TextUtils.isEmpty(clipPath)) {
            SVGAPath path = new SVGAPath();
            path.setValues(clipPath);
            maskPath = path.getPath();
        }
        JSONArray shapes = obj.optJSONArray("shapes");
        if (shapes != null) {
            this.shapes = new SVGAVideoShapeEntity[shapes.length()];
            for (int i = 0; i < shapes.length(); i++) {
                if (shapes.optJSONObject(i) != null) {
                    this.shapes[i] = new SVGAVideoShapeEntity(shapes.optJSONObject(i));
                }
                else {
                    this.shapes[i] = new SVGAVideoShapeEntity();
                }
            }
        }
    }

}

File: library/src/main/java/com/opensource/svgaplayer/SVGAVideoSpriteFrameEntity.kt
package com.opensource.svgaplayer

import android.graphics.Matrix
import android.graphics.Path
import android.text.TextUtils

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.io.Serializable

/**
 * Created by cuiminghui on 2016/10/17.
 */
class SVGAVideoSpriteFrameEntity {

    var alpha: Double
    var layout = SVGARect(0.0, 0.0, 0.0, 0.0)
    var transform = Matrix()
    var maskPath: SVGAPath? = null
    var shapes: List<SVGAVideoShapeEntity> = listOf()

    constructor(obj: JSONObject) {
        this.alpha = obj.optDouble("alpha", 0.0)
        obj.optJSONObject("layout")?.let {
            layout = SVGARect(it.optDouble("x", 0.0), it.optDouble("y", 0.0), it.optDouble("width", 0.0), it.optDouble("height", 0.0))
        }
        obj.optJSONObject("transform")?.let {
            val arr = FloatArray(9)
            val a = it.optDouble("a", 1.0)
            val b = it.optDouble("b", 0.0)
            val c = it.optDouble("c", 0.0)
            val d = it.optDouble("d", 1.0)
            val tx = it.optDouble("tx", 0.0)
            val ty = it.optDouble("ty", 0.0)
            arr[0] = a.toFloat()
            arr[1] = c.toFloat()
            arr[2] = tx.toFloat()
            arr[3] = b.toFloat()
            arr[4] = d.toFloat()
            arr[5] = ty.toFloat()
            arr[6] = 0.0.toFloat()
            arr[7] = 0.0.toFloat()
            arr[8] = 1.0.toFloat()
            transform.setValues(arr)
        }
        obj.optString("clipPath")?.let { d ->
            if (d.isNotEmpty()) {
                maskPath = SVGAPath(d)
            }
        }
        obj.optJSONArray("shapes")?.let {
            val mutableList: MutableList<SVGAVideoShapeEntity> = mutableListOf()
            for (i in 0..it.length() - 1) {
                it.optJSONObject(i)?.let {
                    mutableList.add(SVGAVideoShapeEntity(it))
                }
            }
            shapes = mutableList.toList()
        }
    }

    constructor(obj: ComOpensourceSvgaVideo.FrameEntity) {
        this.alpha = obj.alpha.toDouble()
        if (obj.hasLayout()) {
            obj.layout?.let {
                this.layout = SVGARect(it.x.toDouble(), it.y.toDouble(), it.width.toDouble(), it.height.toDouble())
            }
        }
        if (obj.hasTransform()) {
            obj.transform?.let {
                val arr = FloatArray(9)
                val a = it.a
                val b = it.b
                val c = it.c
                val d = it.d
                val tx = it.tx
                val ty = it.ty
                arr[0] = a
                arr[1] = c
                arr[2] = tx
                arr[3] = b
                arr[4] = d
                arr[5] = ty
                arr[6] = 0.0f
                arr[7] = 0.0f
                arr[8] = 1.0f
                transform.setValues(arr)
            }
        }
        obj.clipPath?.takeIf { it.isNotEmpty() }?.let {
            maskPath = SVGAPath(it)
        }
        this.shapes = obj.shapesList.map {
            return@map SVGAVideoShapeEntity(it)
        }
    }

}
